File: built-in plugins dsp/mb-mallet-glockenspiel.cpp
/**
 * MB Glockenspiel
 * Category : instrument
 * Type     : mallet
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Sparkling metal glockenspiel with bell-like tone
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MALLET_GLOCKENSPIEL_H
#define MB_MALLET_GLOCKENSPIEL_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMalletGlockenspiel : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-mallet-glockenspiel";
    static constexpr const char* PLUGIN_NAME    = "MB Glockenspiel";
    static constexpr const char* PLUGIN_TYPE    = "mallet";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float brightness = 0.9f;  // range [0, 1]
    float sustain_time = 0.6f;  // range [0, 1]
    float mallet = 0.6f;  // range [0, 1]
    float volume = 0.7f;  // range [0, 1]
    };

    MbMalletGlockenspiel() = default;
    ~MbMalletGlockenspiel() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.brightness = std::clamp(params.brightness, 0f, 1f);
        params.sustain_time = std::clamp(params.sustain_time, 0f, 1f);
        params.mallet = std::clamp(params.mallet, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Glockenspiel
        return input;
    }
};

#endif // MB_MALLET_GLOCKENSPIEL_H

File: maxbooster7.5/built-in plugins dsp/mb-mallet-glockenspiel.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbMalletGlockenspielPlugin: PluginDefinition = {
  id: "mb-mallet-glockenspiel",
  slug: "mb-mallet-glockenspiel",
  name: "MB Glockenspiel",
  category: "instrument",
  type: "mallet" as any,
  version: "1.0.0",
  description: "Sparkling metal glockenspiel with bell-like tone",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sine", detune: 0, gain: 0.5 },
    { type: "sine", detune: 2400, gain: 0.3 },
    { type: "sine", detune: 3600, gain: 0.2 },
  ],
  envelope: { attack: 0.001, decay: 1.2, sustain: 0.0, release: 0.8 },
  parameters: [
    {
      id: "brightness",
      name: "Brightness",
      type: "float",
      defaultValue: 0.9,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "sustain_time",
      name: "Sustain",
      type: "float",
      defaultValue: 0.6,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "mallet",
      name: "Mallet",
      type: "float",
      defaultValue: 0.6,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: {
    brightness: 0.9,
    sustain_time: 0.6,
    mallet: 0.6,
    volume: 0.7,
  },
};

export default MbMalletGlockenspielPlugin;
